graphql-kotlin-federation/src/main/kotlin/com/expedia/graphql/federation/FederationSchemaGeneratorHooks.kt

```diff
@@ -0,0 +1,123 @@
+package com.expedia.graphql.federation
+
+import com.expedia.graphql.annotations.GraphQLName
+import com.expedia.graphql.extensions.print
+import com.expedia.graphql.federation.directives.FieldSet
+import com.expedia.graphql.federation.types.ANY_SCALAR_TYPE
+import com.expedia.graphql.federation.types.FIELD_SET_SCALAR_TYPE
+import com.expedia.graphql.federation.types.SERVICE_FIELD_DEFINITION
+import com.expedia.graphql.federation.types._Service
+import com.expedia.graphql.federation.types.generateEntityFieldDefinition
+import com.expedia.graphql.hooks.SchemaGeneratorHooks
+import graphql.TypeResolutionEnvironment
+import graphql.schema.DataFetcher
+import graphql.schema.FieldCoordinates
+import graphql.schema.GraphQLCodeRegistry
+import graphql.schema.GraphQLObjectType
+import graphql.schema.GraphQLSchema
+import graphql.schema.GraphQLType
+import graphql.schema.GraphQLTypeUtil
+import kotlin.reflect.KType
+import kotlin.reflect.full.findAnnotation
+
+open class FederationSchemaGeneratorHooks(private val federatedTypeRegistry: FederatedTypeRegistry) : SchemaGeneratorHooks {
+
+    override fun willGenerateGraphQLType(type: KType): GraphQLType? = when (type.classifier) {
+        FieldSet::class -> FIELD_SET_SCALAR_TYPE
+        else -> null
+    }
+
+    override fun willBuildSchema(builder: GraphQLSchema.Builder): GraphQLSchema.Builder {
+        val originalSchema = builder.build()
+        val originalQuery = originalSchema.queryType
+
+        val federatedSchema = GraphQLSchema.newSchema(originalSchema)
+        val federatedQuery = GraphQLObjectType.newObject(originalQuery)
+        val federatedCodeRegistry = GraphQLCodeRegistry.newCodeRegistry(originalSchema.codeRegistry)
+
+        val entityTypeNames = originalSchema.allTypesAsList
+            .asSequence()
+            .filterIsInstance<GraphQLObjectType>()
+            .filter { type -> type.getDirective("key") != null }
+            .onEach { type -> type.validateFederatedType() }
+            .map { it.name }
+            .toSet()
+
+        // register new federated queries
+        if (entityTypeNames.isNotEmpty()) {
+            federatedQuery.field(SERVICE_FIELD_DEFINITION)
```

Comment:
A service could be federated without exposing any entities, eg to provide new fields on the query or mutation root types.

Reply:
nice catch! thanks